invoices/invoiceregistry.go

```diff
@@ -651,7 +651,24 @@ func (i *InvoiceRegistry) CancelInvoice(payHash lntypes.Hash) error {
 
 	log.Debugf("Invoice(%v): canceling invoice", payHash)
 
-	invoice, err := i.cdb.CancelInvoice(payHash)
+	updateInvoice := func(invoice *channeldb.Invoice) error {
```

Comment:
As explained above, the mutating callback is required to keep the operation atomic. Asserting a pre-condition would also require re-evaluating the updated invoice to see what needs to be done with the newer version of the invoice. Also, checking the pre-condition may not be that easy. It could be that the invoice state enum hasn't changed, but that just an htlc has been added. Also in that case, we need to re-execute the accept/settle logic.

Reply:
>Even further, we wouldn't need Cancel/Settle/Accept calls if all the actual logic is behind a call back like this 

This is indeed what happens in the next commit.

The idea is to have the logic live in a single place. I didn't want to separate the checks from the updates, because it requires a separate data structure to communicate what needs to be updated from invoice registry to channeldb. The check logic gets more complicated with multi-path payments and I don't think that should all be implemented in channeldb. Therefore the decision to move it all into invoice registry.